### PR TITLE
Introducing a functionality to reset the time for produced particles (neutrons)

### PR DESCRIPTION
**Pull request by cfichtlscherer**
Neutrons are produced either by (n,xn)-reactions or fission. At the moment, the time variable p.time() is not reset for the produced neutrons. This is now possible:

`settings.reset_time_for_particle_production = True`

Using this setting in combination with 

`settings.create_delayed_neutrons = False`

OpenMC should now be capable to determine the [prompt neutron lifetime](https://www.nuclear-power.com/nuclear-power/fission/prompt-neutrons/prompt-generation-time-mean-generation-time/) by using a tally with the corresponding scores in combination with a TimeFilter.

See also the following discussions:
https://openmc.discourse.group/t/kinetic-parameter-evaluation-update/302
https://openmc.discourse.group/t/kinetic-parameter-calculation/596

I am looking forward to your feedback.

### CODE REVIEW

**Comment by paulromano** [comment]
@cfichtlscherer A few general questions for you:

1. Do other MC codes have a similar option?
2. Can you elaborate on how the tally with the time filter would work?
3. What application would you use the prompt neutron lifetime for? Isn't is more desirable to have the effective generation time?

**Comment by cfichtlscherer** [comment]
Dear Paul,

Sorry for my late reply. My idea is to allow the user to determine more kinetic parameters to help them understand their system more profoundly. Here are two examples:

**1) Prompt Neutron Generation Time – Mean Generation Time**
_"Prompt Neutron Generation Time (or Mean Generation Time), Λ, is the average time from a prompt neutron emission to a capture that results only in fission."_ 
(https://www.nuclear-power.com/nuclear-power/fission/prompt-neutrons/prompt-generation-time-mean-generation-time/)

My idea is to determine the prompt neutron generation time as the following example:

Change the settings such that only prompt neutrons are modeled and reset the time of the produced neutrons:
```
settings.create_delayed_neutrons = False
settings.reset_time_for_particle_production = True
```

Define a time filter to obtain the time it takes a neutron from its emission till a fission reaction:
```
bin_boarders = list(1e-10 * np.arange(30000)) + [100]	# overflow bin	
tally = openmc.Tally(name="average time till fission")
tally.scores = ['fission']
tally.filters = [openmc.TimeFilter(bin_boarders)]
```

Check that the overflow bin is empty and evaluate:

```
sp = openmc.StatePoint("statepoint.50.h5")
tally = sp.get_tally(name="average time till fission")
nu_fission_per_neutrons = tally.get_values(scores=["nu-fission"]).flatten()
fission_time = tally2.get_values(scores=["fission"]).flatten()[:-1] 
prompt_neutron_generation_time = np.sum(np.asarray(1e-10 * np.arange(30000 - 1) + 5e-11) * fission_time / np.sum(fission_time))     
```

This would provide the user with the prompt neutron generation time.

In the "MCNP Code Version 6.3.0 Theory & User Manual" I can only find (p419):
_"For the point-kinetics parameters, the MCNP code can calculate the following parameters for criticality: the ***neutron generation time*** (Λ), the effective delayed neutron fraction (β eff ), and Rossi-α."_
There is no further information given in the manual. I do not have an MCNP license, so I cannot use this code and try the functionality. In my understanding, the neutron generation time also includes delayed neutrons. This could be the next step, but we would need data on the time distribution of the delayed neutrons.

**2) Prompt Neutron Lifetime**
There are different definitions of the prompt neutron lifetime. The difference relies on including escaping neutrons.

1. _"Prompt neutron lifetime, l, is the average time from a prompt neutron emission to its absorption (fission or radiative capture) or its escape from the system."_
(https://www.nuclear-power.com/nuclear-power/fission/prompt-neutrons/prompt-generation-time-mean-generation-time/)

2. _"The average time between the emission of the prompt neutrons and their absorption in a reactor is called the prompt neutron lifetime and is denoted by l"_
 (Introduction to Nuclear Engineering" by Lamarsh, Baratta (third version) (p330))

The value according to the first definition can be obtained using the `inverse-velocity` score, the second could be obtained (with this pull request) using time filters for determining average absorption times.

I can reproduce the values obtained with the `inverse-velocity` score by tallying the amount of absorbed and escaped neutrons and the average time they need..

I hope this provided a bit more background on the use cases of this functionality. I am happy for all suggestions and feedback.

Christopher

File: src/particle.cpp [hunk 1]
```diff
@@ -77,8 +77,11 @@ void Particle::create_secondary(
   bank.r = r();
   bank.u = u;
   bank.E = settings::run_CE ? E : g();
-  bank.time = time();
-
+  if (settings::reset_time_for_particle_production == true) {
```
Comment (gridley) [inline src/particle.cpp:80]: `if (settings::reset_time_for_particle_production) { ...`

File: src/physics.cpp [hunk 2]
```diff
@@ -205,7 +205,11 @@ void create_fission_sites(Particle& p, int i_nuclide, const Reaction& rx)
     SourceSite site;
     site.r = p.r();
     site.particle = ParticleType::neutron;
-    site.time = p.time();
+    if (settings::reset_time_for_particle_production == true) {
```
Comment (gridley) [inline src/physics.cpp:208]: Same. no need to compare against `true`

**Review by gridley — CHANGES_REQUESTED** [review]
This is interesting, at a minimum academically. I'm pretty curious what the energy spectrum in a reactor would look like as you filter in time. You should see some beautiful hump move from fission spectrum, to intermediate, to a thermal peak as it progresses. I don't know if anyone has created such a visualization before! It should be trivially easy to do after adding this, then doing an eigenvalue mode calculation on a pin cell or something.

My only comment on the code is that you don't need to compare booleans against true.

**Comment by cfichtlscherer** [comment]
@gridley thanks a lot! I just changed the if-conditions

**Comment by paulromano** [comment]
@cfichtlscherer I'm still not totally sure of the utility and generality of this approach. I would recommend that we hold off on this feature for now. I have a project that is starting up in conjunction with @copenhagenatomics that will result in a full capability for generating kinetics parameters. Once that is underway, we should revisit this topic and figure out what your needs are and make sure that we're moving on a path that is agreeable for all.

**Pull request closed without merging**